### PR TITLE
kubeadm: graduate etcd phase

### PR DESCRIPTION
**Pull request by neolit123**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:

move the etcd phase to phase runner.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Refs https://github.com/kubernetes/kubeadm/issues/1163

**Special notes for your reviewer**:
NONE

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

/assign @fabriziopandini 
@kubernetes/sig-cluster-lifecycle-pr-reviews 
/kind feature
/area kubeadm
/hold

depends on:
https://github.com/kubernetes/kubernetes/pull/70331
https://github.com/kubernetes/kubernetes/pull/70098
https://github.com/kubernetes/kubernetes/pull/70097

### CODE REVIEW

**Comment by neolit123** [comment]
@fabriziopandini 
this preserved local etcd as a sub-phase. please let me know if it needs restructure.

File: cmd/kubeadm/app/cmd/phases/etcd.go [hunk 1]
```diff
@@ -17,86 +17,70 @@ limitations under the License.
...
-		use:      "local",
-		short:    "Generates the static Pod manifest file for a local, single-node etcd instance",
-		long:     etcdLocalLongDesc,
-		examples: etcdLocalExample,
-		cmdFunc:  etcdphase.CreateLocalEtcdStaticPodManifestFile,
+// NewEtcdPhase creates a kubeadm workflow phase that implements handling of etcd.
+func NewEtcdPhase() workflow.Phase {
+	phase := workflow.Phase{
+		Name:    "etcd",
+		Short:   "Generates static Pod manifest file for etcd.",
```
Comment (fabriziopandini) [inline cmd/kubeadm/app/cmd/phases/etcd.go]: for local/stacked etcd

File: cmd/kubeadm/app/cmd/phases/etcd.go [hunk 2]
```diff
@@ -17,86 +17,70 @@ limitations under the License.
...
-	cmd := &cobra.Command{
-		Use:     properties.use,
-		Short:   properties.short,
-		Long:    properties.long,
-		Example: properties.examples,
-		Run:     runCmdPhase(properties.cmdFunc, &outDir, &cfgPath, cfg, defaultKubernetesVersion),
+func newEtcdLocalSubPhase() workflow.Phase {
+	phase := workflow.Phase{
+		Name:    "local",
+		Short:   "Generates the static Pod manifest file for a local, single-node etcd instance.",
```
Comment (fabriziopandini) [inline cmd/kubeadm/app/cmd/phases/etcd.go]: for local/stacked etcd instance

File: cmd/kubeadm/app/cmd/phases/etcd.go [hunk 3]
```diff
@@ -17,86 +17,70 @@ limitations under the License.
...
+		}
+		cfg := data.Cfg()
+
+		// Add etcd static pod spec only if external etcd is not configured
+		if cfg.Etcd.External == nil {
+			if err := etcdphase.CreateLocalEtcdStaticPodManifestFile(data.ManifestDir(), cfg); err != nil {
+				return errors.Wrap(err, "error creating local etcd static pod manifest file")
+			}
+		} else {
+			glog.V(1).Infof("[etcd] found external etcd. Skipping the creation of a manifest for local etcd")
```
Comment (fabriziopandini) [inline cmd/kubeadm/app/cmd/phases/etcd.go]: external etcd mode. skipping.... (same message as skipping ects certs generation)

**Review by fabriziopandini — APPROVED** [review]
@neolit123 well done.
Poke me when all the precedence are merged for a final pass

**Comment by neolit123** [comment]
@fabriziopandini updated + rebased.

File: cmd/kubeadm/app/cmd/phases/etcd.go [hunk 4]
```diff
@@ -19,84 +19,71 @@ package phases
...
-		examples: etcdLocalExample,
-		cmdFunc:  etcdphase.CreateLocalEtcdStaticPodManifestFile,
+// NewEtcdPhase creates a kubeadm workflow phase that implements handling of etcd.
+func NewEtcdPhase() workflow.Phase {
+	phase := workflow.Phase{
+		Name:    "etcd",
+		Short:   "Generates static Pod manifest file for local/stacked etcd.",
+		Example: etcdLocalExample,
+		Phases: []workflow.Phase{
+			newEtcdLocalSubPhase(),
```
Comment (fabriziopandini) [inline cmd/kubeadm/app/cmd/phases/etcd.go]: might be a phase with only a subphase is kind of verbose, but I'm fine with this for now...

**Review by fabriziopandini — APPROVED** [review]
@neolit123 Great!
/approve
/lgtm

File: cmd/kubeadm/app/cmd/phases/etcd.go (hunk 4 above)
Comment (neolit123) [reply cmd/kubeadm/app/cmd/phases/etcd.go]: ok, also @timothysc for feedback on this.
LMK if you want to refactor "etcd" into a parent phase that implies "local etcd" only.

Comment (timothysc) [reply cmd/kubeadm/app/cmd/phases/etcd.go]: I'm ok with this for now.

File: cmd/kubeadm/app/cmd/phases/etcd.go [hunk 5]
```diff
@@ -19,84 +19,71 @@ package phases
...
-		use:      "local",
-		short:    "Generates the static Pod manifest file for a local, single-node etcd instance",
-		long:     etcdLocalLongDesc,
-		examples: etcdLocalExample,
-		cmdFunc:  etcdphase.CreateLocalEtcdStaticPodManifestFile,
+// NewEtcdPhase creates a kubeadm workflow phase that implements handling of etcd.
+func NewEtcdPhase() workflow.Phase {
+	phase := workflow.Phase{
+		Name:    "etcd",
+		Short:   "Generates static Pod manifest file for local/stacked etcd.",
```
Comment (timothysc) [inline cmd/kubeadm/app/cmd/phases/etcd.go]: I don't think you need to say "stacked" in multiple locations, I'd probably remove that.

**Review by timothysc — CHANGES_REQUESTED** [review]
remove all the references to "stacked" then LGTM 
/approve
/hold cancel

**Review by timothysc — APPROVED** [review]
/lgtm

**Pull request merged**